lib/jet_ext/ecto/schemaless/schema.ex

```diff
@@ -0,0 +1,140 @@
+defmodule JetExt.Ecto.Schemaless.Schema do
+  @moduledoc false
+
+  use TypedStruct
+
+  @typep options() :: [
+           source: Ecto.Schema.source(),
+           prefix: Ecto.Schema.prefix(),
+           primary_key: [atom()],
+           autogenerate: [autogenerate_option()],
+           autoupdate: [autogenerate_option()],
+           constraints: [Ecto.Changeset.constraint()]
+         ]
+
+  @typep autogenerate_option() ::
+           {fields :: [atom()], generator :: {module(), fun :: atom(), args :: [term()]}}
+
+  @type row() :: map()
+
+  typedstruct do
+    field :types, Ecto.Changeset.types(), enforce: true
+    field :source, Ecto.Schema.source(), enforce: true
+    field :prefix, Ecto.Schema.prefix()
+    field :primary_key, [atom()], enforce: true
+    field :autogenerate, [autogenerate_option()]
+    field :autoupdate, [autogenerate_option()]
+    field :constraints, [Ecto.Changeset.constraint()]
+  end
+
+  @spec new(Ecto.Changeset.types(), options()) :: t()
+  def new(types, options) do
+    source = Keyword.fetch!(options, :source)
+    prefix = Keyword.get(options, :prefix)
+    primary_key = Keyword.fetch!(options, :primary_key)
+    autogenerate = Keyword.get(options, :autogenerate, [])
+    autoupdate = Keyword.get(options, :autoupdate, [])
+    constraints = Keyword.get(options, :constraints, [])
+
+    %__MODULE__{
+      types: types,
+      source: source,
+      prefix: prefix,
+      primary_key: primary_key,
+      autogenerate: autogenerate,
+      autoupdate: autoupdate,
+      constraints: constraints
+    }
+  end
+
+  @spec changeset(t(), data :: map(), params :: map(), permitted :: nil | [atom()]) ::
+          Ecto.Changeset.t(row())
+  def changeset(%__MODULE__{} = schema, data \\ %{}, params, permitted \\ nil) do
+    permitted = permitted || Map.keys(schema.types)
+
+    # https://www.notion.so/byzanteam/Form-Metal-1959b7eda1d34f349d2106a9f9838844?pvs=4#f39c5c2599e8477e9c6fb2049b102a18
+    Ecto.Changeset.cast({data, schema.types}, params, permitted,
+      empty_values: [[] | Ecto.Changeset.empty_values()]
+    )
+  end
```

Comment:
1. 这个函数我觉得不要设计 params 和 permitted 进去吧，这个函数就生成一个 changeset 就好了，剩下的部分可以直接用 Ecto.Changeset.cast 去做
2. 这里的 empty_values 也可以处理成在真正 cast 的时候去处理


```suggestion
  @spec changeset(t(), data :: map()) ::
          Ecto.Changeset.t(row())
  def changeset(%__MODULE__{} = schema, data \\ %{}) do
    Ecto.Changeset.change({data, schema.types})
  end
```